core/cog-prefix-routes-handler.c

```diff
@@ -0,0 +1,341 @@
+ * used to serve requests which do not match any of the routed prefixes.
+ *
+ * The set of available prefix routes can be configured using
+ * [method@Cog.PrefixRoutesHandler.mount] and
+ * [method@Cog.PrefixRoutesHandler.unmount]. For each request, routes
+ * are checked and the one that matches the most URI *path* components
+ * will handle the request.
+ *
+ * This handler is typically used in tandem with
+ * [class@Cog.DirectoryFilesHandler], the latter being typically a
+ * fallback, or as the handler for a routed prefix.
+ *
+ * The following configures a handler to route requests with their path
+ * starting with `/static/images` to a certain directory, and all the
+ * rest of paths to another—note how the
+ * [property@Cog.DirectoryFilesHandler:strip-components] property is
+ * used to remove the matched path components for that route:
+ *
+ * ```c
+ * GFile *content_dir = g_file_new_for_path ("data/content");
+ * CogRequestHandler *fallback_handler = cog_directory_files_handler_new (fallback_dir);
+ *
+ * GFile *images_dir = g_file_new_for_path ("data/images");
+ * CogRequestHandler *images_handler = g_object_new (COG_TYPE_DIRECTORY_FILES_HANDLER,
+ *                                                   "base-path", images_dir,
+ *                                                   "strip-components", 2,
+ *                                                   NULL);
+ *
+ * CogRequestHandler *handler = cog_prefix_routes_handler_new (fallback_handler);
+ * cog_prefix_routes_handler_mount (COG_PREFIX_ROUTES_HANDLER (handler),
+ *                                  "/static/images",
+ *                                  images_handler);
+ * ```
+ */
+
+struct _CogPrefixRoutesHandler {
+    GObject parent;
+
+    CogRequestHandler *fallback;
+    GHashTable        *routes;  /* string -> CogRequestHandler */
+};
+
+enum {
+    PROP_0,
+    PROP_FALLBACK_HANDLER,
+    N_PROPERTIES,
+};
+
+static GParamSpec *s_properties[N_PROPERTIES] = { NULL, };
+
+
+
+static void
+cog_prefix_routes_handler_run_fallback (CogPrefixRoutesHandler *self,
+                                        WebKitURISchemeRequest *request)
+{
+    if (self->fallback) {
+        cog_request_handler_run (self->fallback, request);
+    } else {
+        g_autoptr(GError) error =
+            g_error_new (G_FILE_ERROR,
+                         G_FILE_ERROR_NOENT,
+                         "No file for URI path: %s",
+                         webkit_uri_scheme_request_get_path (request));
+        webkit_uri_scheme_request_finish_error (request, error);
+    }
+}
+
+
+static void
+cog_prefix_routes_handler_run (CogRequestHandler      *request_handler,
+                               WebKitURISchemeRequest *request)
+{
+    CogPrefixRoutesHandler *self = COG_PREFIX_ROUTES_HANDLER (request_handler);
+
+    const char *uri_path = webkit_uri_scheme_request_get_path (request);
+    if (!uri_path)
+        return cog_prefix_routes_handler_run_fallback (self, request);
```

Comment:
if the WebKit can resolve the path, it is not much more simple just end the logic at this point with a `webkit_uri_scheme_request_finish_error()`?

Reply:
The path can be empty because an URI can have an empty path and even an empty authority section, both `foo:` and `foo://authority are valid, and the path will be `NULL` in that case. The `CogPrefixRoutesHandler` cannot handle an empty path because it always needs to match a path prefix, so if there is no path the idea is to pass the request to the fallback handler, which may support URIs with empty paths.

Reply:
Ok. I see the point. Makes sense.